Binary tree

PonizhenDoRyadovogo opened the pull request:
Здравствуйте, доделал лабораторную по Двоичному дереву

Commit:
Создан проект BinaryTree и добавлено:

1. Добавлены классы Node и BinaryTree
2. Сделаны их конструкторы
3. Добавлены методы получения ключа/узла
4. Добавления узла
5. Установки ключа
6. Вывод дерева

Commit:
Произведен рефакторинг функций, функция _AddNode перемещена в private

Commit:
Произведен рефакторинг функций(в прошлый раз изменения в cpp файле не отразились)

Commit:
Обратный рефакторинг функций, добавлен метод clear()

Commit:
Добавлен метод копирования поддеревьев узла, конструктор копирования

Commit:
Добавлены методы проверки на пустоту,подсчет количества узлов в дереве, получение корня

Commit:
Изменения:

1. Добавлена функция подсчета выстоы дерева
2. Функция поиска по ключу
3. Изменена функция clear() и добавлена функция удаления поддеревьев дерева

Commit:
Изменения:

1. Деструктор, функции поиска по ключу, добавления узла сделаны виртуальнми, чтобы в дальнейшем можно было сделать наследование в БД поиска
2. Добавлен оператор =
3. Добавлена приватная функция, которая создает список из дерева
4. Добавлена фукция нахождения "родителя" потомка
5. Добавлена функция удаления узла(рассмотрен только случай, если удаляемый узел - это корень дерева(да, там просто свапаются ключи листа и корня, после этго удаляется лист)

Commit:
Добавлены функции получение всех листов дерева и вывода их на консоль

Commit:
Изменения:

1. Добавил функцию получения вектора, содержащего все ключи по возрастанию
2. Добавлен модификатр const в функциях

Commit:
Изменения:

1. Сделана функция проверки на сбалансированность дерева
2. Добавлены функции, возращающие min/max 
3. Подсчета уровень узла с заданным ключом
4. Добавлена приватная функция, создающая вектор листов от заданного узла
5. Доделана функция remove()

Commit:
Рефакторинг

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.h:
```diff
@@ -0,0 +1,69 @@
+	virtual Node* addNode(const int key);
+	virtual bool remove(const int key);
+	virtual bool remove(Node* root, const int key);
+	virtual Node* parent(const Node* child) const;
```
Слишком много виртуальных функций.
Перечитайте реализации этих методов и подумайте, поведения каких из них в будущем понадобится переопределять.

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+    clear();
+}
+
+void BinaryTree::clearFrom(Node *root)
```
```suggestion
void BinaryTree::clearChildren(Node *root)
```

PonizhenDoRyadovogo replied:
Переименовал функцию

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+    int left = height(root->getLeft());
+    int right = height(root->getRight());
+    int difference = right - left;
+    bool isBalance = false;
+    if (difference == 0 || difference == 1 || difference == -1)
+        isBalance = true;
+    return isBalance * isBalanced(root->getLeft()) * isBalanced(root->getRight());
```
```suggestion
    bool balanced = (difference <= 1 && difference >= -1);
    return balanced && isBalanced(root->getLeft()) && isBalanced(root->getRight());

```

PonizhenDoRyadovogo replied:
исправил

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+        return root;
+    else
+    {
+        Node* left = find(root->getLeft(), key);
+        Node* right = find(root->getRight(), key);
+        if (left)
+            return left;
+        else if (right)
+            return right;
+        else
+            return nullptr;
```
Зачем сразу искать в левом поддереве, если мы уже могли найти узел в правом?
```suggestion
        Node* node = find(root->getLeft(), key);
        if (!node)
            node = find(root->getRight(), key);
        return node;
```

PonizhenDoRyadovogo replied:
сделано

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+        delete replacementNode;
+        return true;
+    }
+    else if (!removableNode->getLeft() && !removableNode->getRight())//no childs
```
А если removableNode == nullptr?

PonizhenDoRyadovogo replied:
Добавил проверку

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+    return false;
+}
+
+bool BinaryTree::remove(Node* root, const int key)
+{
+    return false;
+}
```
?

PonizhenDoRyadovogo replied:
забыл убрать эту функцию. Убрал

tvz-041 commented on 2_BinaryTree/2_BinaryTree/BinaryTree.cpp:
```diff
@@ -0,0 +1,451 @@
+    return vec;
+}
+
+int BinaryTree::min() const
+{
+    std::vector<int> vec = treeToVector();
+    return vec[0];
+}
+
+int BinaryTree::max() const
+{
+    std::vector<int> vec = treeToVector();
+    return vec.back();
+}
```
А если дерево пусто?

PonizhenDoRyadovogo replied:
Добавил проверку, что если дерево пусто, то возвращаю INT_MIN/INT_MAX соответственно

PonizhenDoRyadovogo replied:
также добавил проверку в treeToVector(), если дерево пусто, то возвращаю пустой вектор

Commit:
Рефакторинг функции удаления поддерева

Commit:
Изменения:

1. Добавил проверки в функции min()/max(), теперь если дерево пусто, то функции вернут INT_MIN/INT_MAX соответсвенно
2. Изменил функцию treeToVector(), теперь если дерево пусто, то возвращается пустой вектор
3. небольшие оптимизации функций fin(None* root, const int key) и isBalanced()
4. Добавлена проверка в функцию remove(const int key), если в дереве нет узла с таким ключом

Commit:
Рефакторинг

Commit:
Изменения:

1. Доработана функция remove(const int key). Добавлена проверка на то, что в дереве только корень. И изменен алгоритм, если узел имеет только одного потомка
2. Доработана функция _treeToList
3. Функция _addNode(Node* root, const int key) перемещена в potected и сделана вируальной
4. Функция find(const int key) const; перестала быть виртуальной

Commit:
Изменения:

1. Добавлен метод size() возвращающий количество узлов в дереве
2. Добавлен класс для тестирования Двоичного дерева
